grpc/stream.go

```diff
@@ -22,13 +21,177 @@ var ErrUnverifiedConnection = errors.New("unverified connection")
 // for a connection.
 var ErrNilAuthentication = errors.New("nil authentication")
 
-// StreamService implements the rpc.SmpcServer interface using a gRPC service.
+// safeStream wraps a grpc.Stream and ensures that it is safe for concurrent
+// use. It prevents multiple concurrent writes, and multiple concurrent reads.
+// It allows a concurrent writer and reader.
+type safeStream struct {
+	grpc.Stream
+
+	done   chan struct{}
+	sendMu *sync.Mutex
+	recvMu *sync.Mutex
+}
+
+func newSafeStream(stream grpc.Stream) safeStream {
+	return safeStream{
+		Stream: stream,
+
+		done:   make(chan struct{}),
+		sendMu: new(sync.Mutex),
+		recvMu: new(sync.Mutex),
+	}
+}
+
+// Send implements the stream.Stream interface.
+func (str safeStream) Send(message stream.Message) error {
+	str.sendMu.Lock()
+	defer str.sendMu.Unlock()
+
+	if str.done == nil {
+		return stream.ErrSendOnClosedStream
+	}
```

Comment:
Why not using the common pattern for done channel ? I think this will cause data racing.
```go
select{
case <- stream.done
    return stream.ErrSendOnClosedStream
default:
}
```